File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char
import Data.List
import qualified Data.Text as T

responseFor :: String -> String
responseFor xs
  | question && shouted = "Calm down, I know what I'm doing!"
  | question            = "Sure."
  | shouted             = "Whoa, chill out!"
  | empty               = "Fine. Be that way!"
  | otherwise           = "Whatever."
  where
    trimmed = (T.unpack . T.strip . T.pack) xs
    question = not (null trimmed) && last trimmed == '?'
    shouted = not (null $ find isAlpha xs) && null (find (\c -> isAlpha c && isLower c) xs)
    empty = not $ any (not . isSpace) xs

File: elixir/bob/lib/bob.ex
defmodule Bob do
  @spec hey(String.t()) :: String.t()
  def hey(input) do
    q = String.trim(input)

    is_shouting =
      String.match?(q, ~r/[[:upper:]]/u) and
        String.match?(q, ~r/^([[:upper:]]|[[:digit:]]|[[:space:]]|[[:punct:]])+[?.!]?$/u)

    is_asking = String.at(q, -1) == "?"

    cond do
      q == "" -> "Fine. Be that way!"
      is_shouting and is_asking -> "Calm down, I know what I'm doing!"
      is_shouting -> "Whoa, chill out!"
      is_asking -> "Sure."
      true -> "Whatever."
    end
  end
end
